File: WEEK4/bj3_1717.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const parent = Array.from({ length: n + 1 }, (_, i) => i);

function find(i) {
    if (parent[i] === i) {
        return i;
    }
    return parent[i] = find(parent[i]);
}

function union(a, b) {
    const rootA = find(a);
    const rootB = find(b);
    if (rootA !== rootB) {
        parent[rootB] = rootA;
    }
}

let result = '';
for (let i = 1; i <= m; i++) {
    const [op, a, b] = input[i].split(' ').map(Number);
    if (op === 0) {
        union(a, b);
    } else {
        if (find(a) === find(b)) {
            result += "YES\n";
        } else {
            result += "NO\n";
        }
    }
}

console.log(result.trim());

File: WEEK4/bj3_1717.py
import sys

sys.setrecursionlimit(10**6)
input = sys.stdin.readline


def find(i):
    if parent[i] == i:
        return i
    parent[i] = find(parent[i])
    return parent[i]


def union(a, b):
    rootA = find(a)
    rootB = find(b)
    if rootA != rootB:
        parent[rootB] = rootA


n, m = map(int, input().split())
parent = list(range(n + 1))

for _ in range(m):
    op, a, b = map(int, input().split())
    if op == 0:
        union(a, b)
    else:
        if find(a) == find(b):
            print("YES")
        else:
            print("NO")
